## Переменные. Математические операции с переменными

Коротко о том, что из себя представляет язык Python (или Питон)

Мы используем ресурс Google Collab для того, чтобы сделать ваши первые шаги в языке Пайтон более приятными и доступными. 

Вообще, в рамках курса мы работаем с кодом локально, т.е устанавливая все необходимые программы на компьютер. Это обязательно нужно будет сделать, если вы хотите успешной пройти курс и освоить программирование. Однако скачивание и установка необходимого ПО может занять время и силы. Поэтому тем, кто этого ещё не сделал, я предлагаю воспользоваться этим сервисом для того, чтобы в начале сосредоточиться на изучении языка, а не на решении смежных вопросов. 

В правом верхнем углу страницы вы можете найти кнопку **+ Code**
Используйте её для создания новой ячейки. Вводите или копируйте код, который мы разбираем на вебинаре и нажимайте на кнопку запуска слева.   


Первая программа, которую мы напишем, будет по традиции называтся **Hello World**. Это самый первый шаг, с которого начинается погружение в мир любого языка программирования. Питон не исключение. Более того, если до этого вы уже писали на каком-то другом ЯП кроме питона, то этот прекрасный змей может приятно удивить вас своей элегантностью и простотой, с которой можно запускать программы и писать код. Приступим:

In [2]:
print('Hello World!')

Hello World!


Легко и просто мы вывели на экран нашу первую программу. Сначала мы написали ключевое слово print, которое говорит компьютеру, что сейчас мы будем что-то выводить на экран. А внутри скобок, этим что-то стала строка Hello World!
Важный момент! Мы поместили нашу строку в одинарные кавычки. Почему? Потому что если не помещать строки в кавычки, то компьютер будет думать, что это переменная. Тогда он начнет ругатся и выдавать ошибку, потому что такой переменной нет. Впрочем, до переменных мы ещё обязательно дойдем. А сначала займемся простой математикой

Если  нужно  перемножить  два  числа,  к  примеру  узнать,  сколько  будет  8 × 3,57, мы обычно пользуемся калькулятором либо берем ручку и умно-жаем в столбик на листе бумаги. А что если использовать для подсчетов  Python

In [3]:
8*3.57

28.56

Обратите  внимание,  что  при  записи  числа  3,57  используется  не  за-пятая,  а  точка.  Кроме  того,  в  Python  числа  перемножаются  с  помощью  звездочки (*), а не знака умножения (×).


Представьте, что вы рыли яму и случайно нашли кошелек с 20 золо-тыми монетами. На следующий день вы тихонько залезли в подвал, где
стоит изобретение вашего дедушки —  работающий на паровом ходу ме-ханизм для копирования предметов, и, на ваше счастье, в него удалось запихнуть все 20 монет. Раздался свист, потом щелчок, и устройство вы-дало еще 10 новеньких монеток.

Сколько монет вы накопите, если будете проделывать эту операцию каждый день в течение года? 
На бумаге эти расчеты выглядят примерно так:


*   10 × 365 = 3650
*   20 + 3650 = 3670




In [4]:
10 * 365

3650

In [5]:
20+3650

3670

Но что если о вашем богатстве узнает пронырливая ворона? 
Предположим, она будет каждую неделю залетать в окно и красть по три монетки.

Сколько у вас будет монет через год? В оболочке Python эти расчеты будут выглядеть так:



In [6]:
3*52

156

In [7]:
3670-156

3514

### Операторы в Python

Питон может складывать, вычитать, умножать, делить, возводить в степень числа. И это не полный список того, что он может, просто эти действия самые простые, поэтому давайте разберемся, как их выполнять 

<table class="src" border="1">
<tbody>
<tr style="background-color: #2b5b84;"><th style="width: 10%;"><span style="color: #ffff00;">Оператор</span></th><th style="width: 45%;"><span style="color: #ffff00;">Описание</span></th><th><span style="color: #ffff00;">Примеры</span></th></tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">+</span></strong></td>
<td>Сложение - Суммирует значения слева и справа от оператора</td>
<td>
<p>15 + 5 в результате будет 20<br> 20 + -3 в результате будет 17<br> 13.4 + 7 в результате будет 20.4</p>
</td>
</tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">-</span></strong></td>
<td>Вычитание - Вычитает правый операнд из левого</td>
<td>15 - 5 в результате будет 10<br> 20 - -3 в результате будет 23<br> 13.4 - 7 в результате будет 6.4</td>
</tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">*</span></strong></td>
<td>Умножение - Перемножает операнды</td>
<td>5 * 5 в результате будет 25<br>7 * 3.2 в результате будет 22.4<br>-3 * 12 в результате будет -36</td>
</tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">/</span></strong></td>
<td>Деление - Делит левый операнд на правый</td>
<td>15 / 5 в результате будет 3<br>5 / 2 в результате будет 2 (В Python 2.x версии при делении двух целых чисел результат будет целое число)<br> 5.0 / 2 в результате будет 2.5 (Чтобы получить "правильный" результат хотя бы один операнд должен быть float)</td>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">**</span></strong></td>
<td>Возведение в степень - возводит левый операнд в степень правого</td>
<td>5 ** 2 в результате будет 25<br>2 ** 3 в результате будет 8<br>-3 ** 2 в результате будет -9</td>
</tr>
</tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">%</span></strong></td>
<td>Деление по модулю - Делит левый операнд на правый и возвращает остаток.</td>
<td>6 % 2 в результате будет 0<br>7 % 2 в результате будет 1<br>13.2 % 5 в результате 3.2</td>
</tr>
<tr>
<td style="background-color: #2b5b84; text-align: center;"><strong><span style="color: #ffff00;">//</span></strong></td>
<td>Целочисленное деление - Деление в котором возвращается только целая часть результата. Часть после запятой отбрасывается.</td>
<td>12 // 5 в результате будет 2<br>4 // 3 в результате будет 1<br>25 // 6 в результате будет 4</td>
</tr>
</tbody>
</table>

Оператор **Деление по модулю** очень интересный. Его часто используют программисты, чтобы узнать четное число или нечетное. 

У четного при делении по модулю на 2 результат будет 0, а у нечетного 1



Порядок выполнения операций

Операции —  это любые действия, которые совершаются с помощью операторов. 
Математические операции выполняются по очереди в зависимости от их приоритета (если не задать другую очередность с помощью скобок). Умножение  и  деление  имеют  более  высокий  приоритет,  чем  сложение  и  вычитание,  и  это  значит,  что  они  будут  выполняться  первыми.  Иначе  говоря, при вычислении математического выражения Python сначала умножит и разделит числа, а затем перейдет к сложению и вычитанию. 

Например,  в  этом  выражении  сперва  будут  перемножены  числа  30 и 20, а затем к их произведению будет прибавлено число 5

In [8]:
5 + 30 * 20

605

По  сути  это  выражение  означает  «умножить  30  на  20  и  прибавить  к  результату  5».  Получается  605.  Однако  мы  можем  изменить  порядок  операций, заключив первые два числа в скобки. Вот так:

In [9]:
(5 + 30) * 20

700

В  результате  получилось  700,  а  не  605,  поскольку  Python  выполняет операции в скобках прежде, чем операции вне скобок. 

Другими словами, это выражение означает «прибавить 5 к 30 и умножить результат на 20».

Скобки могут быть вложенными, то есть внутри скобок могут стоять еще одни скобки:

In [10]:
((5 + 30) * 20) / 10

70.0

In [11]:
# а теперь посмотрим, как это будет выглядеть без скобок

5 + 30 * 20 / 10

65.0

Запомните, что умножение и деление всегда выполняются прежде, чем сложение и вычитание, если не менять порядок вычислений с помощью скобок.

### Решение задач

1) Найти ответ 

![alt text](https://lms.skillfactory.ru/assets/courseware/v1/bd66d9520e44205e024494fe26c1f55b/asset-v1:SkillFactory+PY-34+19FEB2020+type@asset+block/Screenshot_2019-04-09_at_16.42.52.png)

In [27]:
# Важно правильно поставить скобки! А ещё нужно вспомнить, что такое квадратный корень и как его представить в виде степени

(4 + (5*60) - 17 + (65/4))**0.5

#Питон имеет специальную функцию для округления дробных чисел: round(число, количество цифр)

# round((4 + (5*60) - 17 + (65/4))**0.5, 3)

17.414074767267998

![alt text](https://lms.skillfactory.ru/assets/courseware/v1/4e5dd3e701139b139c0fd28e8c7207a8/asset-v1:SkillFactory+PY-34+19FEB2020+type@asset+block/Screenshot_2019-04-09_at_16.59.52.png)

In [28]:
((((9**0.5)**3)**2)**2)/(-2**-3)

-4251528.0

### Переменные

В программировании слово переменная обозначает именованное место для  хранения  данных,  например  чисел,  текста,  списков  с  числами  или  символами и так далее. Также переменную можно рассматривать как ярлык, которым помечены некие данные.

Например, на нашем складе лежит 500 пар кроссовок Nike. Давайте как-то зафиксируем эту информацию!

In [0]:
nikes = 500

Чтобы создать переменную с именем **nikes**, нужно указать имя, поставить знак «равно» (=) и ввести соответствующие данные. Давайте создадим переменную nikes, указав, что ей соответствует значение 500 (однако из этого не следует, что другая переменная не может иметь такое же значение):

In [0]:
найки = 500

Чтобы напечатать значение нашей переменной, введите в оболочке Python  команду  print  и  следом  за  ней  —    имя  переменной  в  скобках.  Вот так:

In [14]:
print(найки)

500


Можно изменить значение переменной `nikes` —  сделать так, чтобы ей соответствовалидругие данные. Например, вот как заменить значение `nikes` числом 450

In [24]:
nikes = 450
print(nikes)

450


### Немного об именах переменных

Для создания корректных имен перемен­ных надо следовать всего нескольким правилам; о некорректном имени Python вам сообщит, выведя ошибку

Правила для создания корректных имен переменных:

Правила наименования переменных

Есть несколько правил при определении переменных:

            имена переменных не должны повторять уже встроенные в Python имена, например, print, str, in и т.д.;
            имена переменных могут содержать только буквы, цифры и нижнее подчёркивание; 
            имена переменных не должны начинаться с цифры;
            очень желательно, чтобы имена переменных выражали суть переменной и не были слишком длинными.

Если компьютер принял переменную это не значит, что это имя хорошее! 
Переменная с именем **aklkjkll112312** будет понятна компьютеру, но ну будет понятна человеку. Поэтом при создании переменных вы должны ориентироваться на своего читателя. Порой вы будете единственным читателем вашего кода, но поверьте(!), вернувшись через месяц для того, чтобы немного изменить свою программу, вы ужаснетесь, насколько там все непонятно! Поэтому, если вы с любовью относитесь к себе и своим коллегам, не пренебрегайте правилами и договоренностями. Они значительно упрощают вашу жизнь. Подробнее про имена переменных тут
http://pythonlearn.ru/python-dlya-nachinayushhix/imena-dlya-peremennyx-v-python/







Итак, у нас есть 500 кроссовок фирмы найк. Но ведь есть же и другие бренды, правильно? Пускай у нас ещё будут кроссовки adidas в количестве 400 штук. 

In [0]:
adidas = 400

Теперь применим математические операции, которые мы изучили на наших переменных

1) Мы можем посчитать общее количество кроссовок, которое у нас есть

2) Мы можем посчитать выручку, которую мы поулучим, реализовав наш товар по стоимости 5 тысяч и 4 тысячи рублей за пару

3) Мы можем узнать сколько мы получим прибыли после уплаты налога, который пусть будет 13%

In [17]:
nikes + adidas

900

In [0]:
raw_income = (nikes * 5000) + (adidas*4000) 

In [19]:
after_nalog = raw_income * (1- 0.13)
print(after_nalog)

3567000.0


In [20]:
after_nalog = raw_income - (raw_income * 0.13)
after_nalog

3567000.0

Давайте введем небольшую вариативность в нашу программу. У функции `print()` есть функция-антагонист, которая позволяет вводить данные в программу с клавиатуры? Она называется `input()`. Но перед этим важно упопянуть такую особенность языка питон как **типизация**. У данных есть типы. Мы работали пока только со строками и числами, но типов данных гораздо больше. Компьютер не может интуитивно понять, что мы от него хотим в отличие от человека. Поэтому важно следить за типами данных, которые мы предоставляем нашей программе. Числа и строки это разные типы данных и работаем мы с ними по-разному.

In [21]:
'5' == 5 # == - это знак "равно". Один = обозначает присваивание переменной, поэтому чтобы проверять равенство используется ==

False

Поэтому чтобы превратить строку в число, мы можем использовать особую функцию `int()` в которую мы передадим строку, а на выходе получим число. Передадим означает, что мы поместим эту строку внутрь скобок.

In [22]:
int('5') == 5 # Ура, теперь верно!

True

In [23]:
nalog = int(input('Укажите налоговую ставку в процентах: '))

after_nalog = raw_income - (raw_income * nalog/100)
after_nalog

3198000.0

Дело в том, что функция `input()` возвращает нам строку, поэтому нам необходимо её обернуть в другую функцию - `int()`, чтобы она стала числом. 

Если вы хотите понять, какой тип переменной перед вами, вы можете воспользоваться встроенной функцией type(). 

In [32]:
type('Привет')

str

In [33]:
type(4)

int

In [34]:
type(4.0)

float

In [35]:
4.0 == 4

True

Важно держать в голове следующие правила работы с типами:

            если хотя бы одна из переменных принадлежит к типу float, результат вычисления тоже получит тип float; 
            при операции деления результат всегда вернётся с типом float, даже если оба исходных числа принадлежат к типу int. 

К строковым (текстовым) переменным, как ни странно, тоже применимы некоторые из изученных операторов — это операторы + и *.  Так, оператор + «склеит» воедино две строки, а оператор * продублирует строку нужное количество раз, объединив результат в одну строку.

Давайте решим какую-нибудь задачу, в которой нам пригодится `input()`

![alt text](https://lms.skillfactory.ru/assets/courseware/v1/422d0ebb90f89a15c0e4f3d573f27f69/asset-v1:SkillFactory+PY-34+19FEB2020+type@asset+block/Screenshot_2019-04-09_at_22.58.17.png)

In [29]:
x = int(input('Введите X: '))

6.996 + 45 * (x/2)**2 + 2*x

Введите X: 100500


113628013506.996

Напишите программу, которая принимает на вход строку и целое число и выводит на экран эту строку, её длину и её длину в степени введённого целого числа. При введении, например, строки "привет" и числа 2 на экране должно быть выведено "привет 6 36". Названия переменных string и power. Для теста используйте значения string = 'Привет!', power = 5

In [31]:
text = input('Введите текст: ')

power = int(input('Введите степень: '))

len_text = len(text)

result = len_text**power

print(text, ' Степень: ', power, ' Результат: ', result)

Введите текст: екекек
Введите степень: 10
екекек  Степень:  10  Результат:  60466176


# Спасибо за внимание!

Буду рад ответить на ваши вопросы и обсудить проблемы

В вебинаре были взяты примеры из книги **Джейсона Бриггса "Python для детей"**  

Также дял изучения основ питона и знакомства прикладными возможностями языка рекомендую книгу **Автоматизация рутинных задач с помощью Python. Эл Свейгарт**